### PR TITLE
Agent class

### PR DESCRIPTION
**Pull request by cameronfabbri**
Class for an Agent. 

To test: `python -m examples.agent_example`

A good question to ask for testing is `"What were the last 10 checking account transactions for Constance Faust?"`

### Prompt
The `Prompt` class implements a consistent interface for constructing a prompt. It can contain `commands`, `examples`, `function_resources`, `data_resources`, `constraints`, and `completed_steps`.

### Agent
The Agent takes as input an `AgentName` for the Agent, a `system_prompt` that acts as a set of instructions, a list of `resources` (paths to SQLite databases) and a list of `modules` (python modules) which act as tools. The Agent first creates a workspace to save data to at `.agents/AgentName/`. The `resources` and `modules` are summarized and saved to `.agents/AgentName/resources.yaml` and `.agents/AgentName/tools.yaml` respectively. When a user gives a command, these summaries are inserted into the prompt such that the Agent can draw from them as needed.

The `UNLQ_GPT_BASE_PROMPT` in `src/neosophia/agents/system_prompts.py` instructs the agent to operate in a step by step fashion. This is important because a function that was executed in a certain step can return data that will be used in subsequent steps. The data that is returned from function calls is saved in `function_resources` and is added to the prompt. When the Agent generates a step, it outputs `Thoughts`, `Resource`, `Function`, `Parameters` and `Returned`. This is all added to the prompt under the `completed_steps` section, so the Agent is able to determine what's already been completed.

When the Agent determines it has enough information to answer the question, it will call `extract_answer`.

### Improvements
- Only insert tools and resources needed for the current task into the prompt instead of everything
- Summarize previous steps taken to reduce token count
- Be able to insert a command after a step has been taken to fix or change something
- Modify agent such that it's easily added into a Gradio app

Right now it's printing the generated prompt and the response from the LLM — we can change that as needed.

### CODE REVIEW

**Review by bdzimmer — COMMENTED** [review]
A bunch of comments, mostly random stuff.

Would like to see more complete type signatures / docstrings, as well as some unit tests here and there.

File: src/neosophia/agents/base.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+
+                    # Add variable to function resources
+                    function_resources[return_name] = res
+
+                    prompt.add_function_resources(return_name, str(res))
+                    prompt.add_completed_step(response)
+                    prompt_str = prompt.generate_prompt()
+
+                    # Get an approximate token count without needing to encode
+                    num_tokens = len(prompt_str.replace(' ', '')) / 4
```
[the remark this answers is not in the document]
Comment (cameronfabbri) [reply src/neosophia/agents/base.py:268]: It's just for efficiency. Encoding is pretty quick but still takes a lot longer than the approximation, so I figured if we're nowhere near the token limit (from the approximation) then there's no need to encode and check.

**Comment by jmpounders** [comment]
This is very good work.  It's a really nice demonstration of a very general agent built on a broad foundation.

I think some of the elements, like resources and tools, could be factored out into their own first-class citizens.  That's not trivial work, though, so we should tackle those under separate efforts.

I left a few other comments (some of which i now see listed under "improvements" above) and agree with Ben's comments.

File: src/neosophia/agents/base.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
+""" Agent base class """
+import os
+import re
+import types
+import readline
```
[the remark this answers is not in the document]
Comment (cameronfabbri) [reply src/neosophia/agents/base.py:5]: It actually is being used just by importing it. It modifies the `input()` function so you can use the arrow keys to move the cursor around when typing a question

File: src/neosophia/agents/base.py [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        self.tools, self.tools_file = autils.setup_and_load_yaml(
+            self.workspace, 'tools.yaml', 'functions')
+
+        # Go through each database file and get a description from the schema
+        for resource in resources:
+
+            # Only do it for databases we haven't saved in the yaml file yet
+            if resource not in self.resources:
+                print(f'Adding new resource {resource}...')
+                description = autils.get_database_description(resource)
```
[the remark this answers is not in the document]
Comment (cameronfabbri) [reply src/neosophia/agents/base.py:65]: Yes that's correct

File: src/neosophia/agents/base.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+        def get_command(prompt):
+            print('\nAsk a question')
+            user_input = ''
+            while user_input == '':
+                user_input = input('> ')
+            prompt.add_command(user_input)
+            return user_input
+
+        while True:
```
[the remark this answers is not in the document]
Comment (cameronfabbri) [reply src/neosophia/agents/base.py:120]: Just added the ability to exit the program. You can use natural language like "I'm done now" or "I don't have any more questions" 😄

File: src/neosophia/agents/utils.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+def build_function_dict_from_modules(
+        modules: List[types.ModuleType]) -> Dict[str, Tuple[Callable, str]]:
+    """
+    Takes a list of python modules as input and builds a dictionary containing
+    the function name as the key and a tuple containing the callable function
+    and its string representation as the value.
+    """
+
+    function_dict = {}
+    for module in modules:
```
[the remark this answers is not in the document]
Comment (jmpounders) [reply src/neosophia/agents/utils.py:29]: Let's push this to a future effort.

**Review by jmpounders — APPROVED** [review]
As discussed, this is a great start to some powerful capabilities!  This is good to merge as far as I'm concerned.  Once merged we can come up with a plan for what we want to refine and/or split out into independent components.  Thanks!

File: src/neosophia/agents/base.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bdzimmer) [reply src/neosophia/agents/base.py:67]: @jmpounders In the experimentation that Cam and I did with the in memory join capability we started implementing adding custom resources / descriptions (specifically tools).

File: src/neosophia/agents/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bdzimmer) [reply src/neosophia/agents/utils.py:29]: @jmpounders I agree, this approach is convenient for testing but you would need the capability to expose / hide certain functions.

**Review by bdzimmer — APPROVED** [review]

**Pull request merged**